// @anchor-protocol/anchor.js/dist/queries/money-market/liquidation-liquidation-amount.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.queryLiquidationLiquidationAmount = void 0;
const queryLiquidationLiquidationAmount = ({ lcd, borrow_amount, borrow_limit, collaterals, collateral_prices, }) => (addressProvider) => __awaiter(void 0, void 0, void 0, function* () {
    const liquidationContractAddress = addressProvider.liquidation();
    const response = yield lcd.wasm.contractQuery(liquidationContractAddress, {
        liquidation_amount: {
            borrow_amount: borrow_amount,
            borrow_limit: borrow_limit,
            collaterals: collaterals,
            collateral_prices: collateral_prices,
        },
    });
    return response;
});
exports.queryLiquidationLiquidationAmount = queryLiquidationLiquidationAmount;
