// @anchor-protocol/anchor.js/dist/fabricators/basset/basset-update-params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricatebAssetUpdateParams = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const number_1 = require("../../utils/validation/number");
const true_1 = require("../../utils/validation/true");
const fabricatebAssetUpdateParams = ({ address, epoch_period, unbonding_period, peg_recovery_fee, er_threshold, }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        epoch_period ? (0, number_1.validateIsNumber)(epoch_period) : true_1.validateTrue,
        unbonding_period ? (0, number_1.validateIsNumber)(unbonding_period) : true_1.validateTrue,
        //TODO: validate decimal and denom
    ]);
    const bAssetContractAddress = addressProvider.bLunaHub();
    return [
        new terra_js_1.MsgExecuteContract(address, bAssetContractAddress, {
            update_params: {
                epoch_period: epoch_period,
                unbonding_period: unbonding_period,
                peg_recovery_fee: peg_recovery_fee,
                er_threshold: er_threshold,
            },
        }),
    ];
};
exports.fabricatebAssetUpdateParams = fabricatebAssetUpdateParams;
