// @anchor-protocol/anchor.js/dist/fabricators/basset/basset-unbond.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricatebAssetUnbond = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const number_1 = require("../../utils/validation/number");
const create_hook_msg_1 = require("../../utils/cw20/create-hook-msg");
const fabricatebAssetUnbond = ({ address, amount }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, number_1.validateIsNumber)(amount),
        (0, number_1.validateIsGreaterThanZero)(amount),
    ]);
    const bAssetTokenAddress = addressProvider.bLunaToken();
    const bAssetHubAddress = addressProvider.bLunaHub();
    return [
        new terra_js_1.MsgExecuteContract(address, bAssetTokenAddress, {
            // @see https://github.com/Anchor-Protocol/anchor-bAsset-contracts/blob/cce41e707c67ee2852c4929e17fb1472dbd2aa35/contracts/anchor_basset_token/src/handler.rs#L101
            send: {
                contract: bAssetHubAddress,
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: (0, create_hook_msg_1.createHookMsg)({
                    unbond: {},
                }),
            },
        }),
    ];
};
exports.fabricatebAssetUnbond = fabricatebAssetUnbond;
