// @anchor-protocol/anchor.js/dist/fabricators/basset/basset-update-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricatebAssetUpdateConfig = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const true_1 = require("../../utils/validation/true");
const fabricatebAssetUpdateConfig = ({ address, owner, reward_contract, token_contract, airdrop_registry_contract, }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        reward_contract ? (0, address_1.validateAddress)(reward_contract) : true_1.validateTrue,
        token_contract ? (0, address_1.validateAddress)(token_contract) : true_1.validateTrue,
    ]);
    const bAssetContractAddress = addressProvider.bLunaHub();
    return [
        new terra_js_1.MsgExecuteContract(address, bAssetContractAddress, {
            update_config: {
                owner: owner,
                reward_contract: reward_contract,
                token_contract: token_contract,
                airdrop_registry_contract: airdrop_registry_contract,
            },
        }),
    ];
};
exports.fabricatebAssetUpdateConfig = fabricatebAssetUpdateConfig;
