// @anchor-protocol/anchor.js/dist/fabricators/basset/basset-send-from.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricatebAssetSendFrom = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const number_1 = require("../../utils/validation/number");
const fabricatebAssetSendFrom = ({ address, amount, contract, owner, msg }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, number_1.validateIsNumber)(amount),
        (0, number_1.validateIsGreaterThanZero)(amount),
        (0, address_1.validateAddress)(owner),
        (0, address_1.validateAddress)(contract),
    ]);
    const bAssetTokenAddress = addressProvider.bLunaToken();
    let message = undefined;
    if (msg) {
        message = Buffer.from(JSON.stringify(msg)).toString('base64');
    }
    return [
        new terra_js_1.MsgExecuteContract(address, bAssetTokenAddress, {
            // @see https://github.com/Anchor-Protocol/anchor-bAsset-contracts/blob/cce41e707c67ee2852c4929e17fb1472dbd2aa35/contracts/anchor_basset_token/src/handler.rs#L203
            send_from: {
                owner: owner,
                contract: contract,
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: message,
            },
        }),
    ];
};
exports.fabricatebAssetSendFrom = fabricatebAssetSendFrom;
