// @anchor-protocol/anchor.js/dist/__tests__/anchor-beth-test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const test_fabricator_1 = require("../utils/test-fabricators/test-fabricator");
const fabricators_1 = require("../fabricators");
const terra_js_1 = require("@terra-money/terra.js");
const create_hook_msg_1 = require("../utils/cw20/create-hook-msg");
const common_1 = require("../__tests__/common");
const fabricators_2 = require("../../src/fabricators");
/* eslint-disable */
describe('bEth', () => {
    it('update-config', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_2.fabricatebEthUpdateConfig, {
            address: 'address',
            owner: 'new-owner',
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthReward(), {
                update_config: {
                    owner: 'new-owner',
                },
            }),
        ]);
    }));
    it('claim-rewards', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_2.fabricatebEthClaimRewards, {
            address: 'address',
            recipient: undefined,
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthReward(), {
                claim_rewards: { recipient: undefined },
            }),
        ]);
    }));
    it('transfer', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthTransfer, {
            address: 'address',
            amount: '1000',
            recipient: 'recipient',
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                transfer: {
                    recipient: 'recipient',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('transfer-from', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthTransferFrom, {
            address: 'address',
            owner: 'owner',
            amount: '1000',
            recipient: 'recipient',
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                transfer_from: {
                    owner: 'owner',
                    recipient: 'recipient',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('burn', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthBurn, {
            address: 'address',
            amount: '1000',
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                burn: {
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('burn-from', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthBurnFrom, {
            address: 'address',
            owner: 'owner',
            amount: '1000',
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                burn_from: {
                    owner: 'owner',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('send', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthSend, {
            address: 'address',
            amount: '1000',
            contract: 'contract',
            msg: { msg: {} },
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                send: {
                    contract: 'contract',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    msg: (0, create_hook_msg_1.createHookMsg)({ msg: {} }),
                },
            }),
        ]);
    }));
    it('send-from', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthSendFrom, {
            address: 'address',
            amount: '1000',
            owner: 'owner',
            contract: 'contract',
            msg: { msg: {} },
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                send_from: {
                    owner: 'owner',
                    contract: 'contract',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    msg: (0, create_hook_msg_1.createHookMsg)({ msg: {} }),
                },
            }),
        ]);
    }));
    it('increase-allowance', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthIncreaseAllowance, {
            address: 'address',
            amount: '1000',
            spender: 'spender',
            expires: { never: {} },
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                increase_allowance: {
                    spender: 'spender',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    expires: { never: {} },
                },
            }),
        ]);
    }));
    it('decrease-allowance', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testFabricator)(expect, fabricators_1.fabricatebEthDecreaseAllowance, {
            address: 'address',
            amount: '1000',
            spender: 'spender',
            expires: { never: {} },
        }, common_1.addressProvider, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.bEthToken(), {
                decrease_allowance: {
                    spender: 'spender',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    expires: { never: {} },
                },
            }),
        ]);
    }));
});
