// @anchor-protocol/anchor.js/dist/__tests__/cw20-test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const test_fabricator_1 = require("../utils/test-fabricators/test-fabricator");
const fabricators_1 = require("../fabricators");
const common_1 = require("../__tests__/common");
const terra_js_1 = require("@terra-money/terra.js");
const create_hook_msg_1 = require("../utils/cw20/create-hook-msg");
/* eslint-disable */
describe('cw20', () => {
    it('transfer', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20Transfer, {
            address: 'address',
            contract_address: common_1.addressProvider.ANC(),
            amount: '1000',
            recipient: 'recipient',
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                transfer: {
                    recipient: 'recipient',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('transfer-from', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20TransferFrom, {
            address: 'address',
            contract_address: common_1.addressProvider.ANC(),
            owner: 'owner',
            amount: '1000',
            recipient: 'recipient',
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                transfer_from: {
                    owner: 'owner',
                    recipient: 'recipient',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('burn', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20Burn, {
            address: 'address',
            contract_address: common_1.addressProvider.ANC(),
            amount: '1000',
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                burn: {
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('burn-from', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20BurnFrom, {
            address: 'address',
            contract_address: common_1.addressProvider.ANC(),
            owner: 'owner',
            amount: '1000',
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                burn_from: {
                    owner: 'owner',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                },
            }),
        ]);
    }));
    it('send', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20Send, {
            address: 'address',
            amount: '1000',
            contract_address: common_1.addressProvider.ANC(),
            contract: 'contract',
            msg: { msg: {} },
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                send: {
                    contract: 'contract',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    msg: (0, create_hook_msg_1.createHookMsg)({ msg: {} }),
                },
            }),
        ]);
    }));
    it('send-from', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20SendFrom, {
            address: 'address',
            amount: '1000',
            contract_address: common_1.addressProvider.ANC(),
            owner: 'owner',
            contract: 'contract',
            msg: { msg: {} },
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                send_from: {
                    owner: 'owner',
                    contract: 'contract',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    msg: (0, create_hook_msg_1.createHookMsg)({ msg: {} }),
                },
            }),
        ]);
    }));
    it('increase-allowance', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20IncreaseAllowance, {
            address: 'address',
            contract_address: common_1.addressProvider.ANC(),
            amount: '1000',
            spender: 'spender',
            expires: { never: {} },
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                increase_allowance: {
                    spender: 'spender',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    expires: { never: {} },
                },
            }),
        ]);
    }));
    it('decrease-allowance', () => __awaiter(void 0, void 0, void 0, function* () {
        (0, test_fabricator_1.testCw20Fabricator)(expect, fabricators_1.fabricateCw20DecreaseAllowance, {
            address: 'address',
            amount: '1000',
            contract_address: common_1.addressProvider.ANC(),
            spender: 'spender',
            expires: { never: {} },
        }, [
            new terra_js_1.MsgExecuteContract('address', common_1.addressProvider.ANC(), {
                decrease_allowance: {
                    spender: 'spender',
                    amount: new terra_js_1.Int(new terra_js_1.Dec('1000').mul(1000000)).toString(),
                    expires: { never: {} },
                },
            }),
        ]);
    }));
});
