// @anchor-protocol/anchor.js/dist/fabricators/exchange/swap-ust.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateExchangeSwapUSTANC = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const number_1 = require("../../utils/validation/number");
const fabricateExchangeSwapUSTANC = ({ address, amount, to, belief_price, max_spread, denom }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, number_1.validateIsNumber)(amount),
        (0, number_1.validateIsGreaterThanZero)(amount),
    ]);
    const coins = new terra_js_1.Coins([
        new terra_js_1.Coin(denom, new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString()),
    ]);
    const pairAddress = addressProvider.ancUstPair();
    return [
        new terra_js_1.MsgExecuteContract(address, pairAddress, {
            swap: {
                offer_asset: {
                    info: {
                        native_token: {
                            denom: denom,
                        },
                    },
                    amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                },
                belief_price: belief_price,
                max_spread: max_spread,
                to: to,
            },
        }, coins),
    ];
};
exports.fabricateExchangeSwapUSTANC = fabricateExchangeSwapUSTANC;
