// @anchor-protocol/anchor.js/dist/fabricators/exchange/swap-anc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateExchangeSwapANC = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const number_1 = require("../../utils/validation/number");
const create_hook_msg_1 = require("../../utils/cw20/create-hook-msg");
const fabricateExchangeSwapANC = ({ address, amount, to, belief_price, max_spread }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, number_1.validateIsNumber)(amount),
        (0, number_1.validateIsGreaterThanZero)(amount),
    ]);
    const ancTokenAddress = addressProvider.ANC();
    const pairAddress = addressProvider.ancUstPair();
    return [
        new terra_js_1.MsgExecuteContract(address, ancTokenAddress, {
            send: {
                contract: pairAddress,
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: (0, create_hook_msg_1.createHookMsg)({
                    swap: {
                        belief_price: belief_price,
                        max_spread: max_spread,
                        to: to,
                    },
                }),
            },
        }),
    ];
};
exports.fabricateExchangeSwapANC = fabricateExchangeSwapANC;
