// @anchor-protocol/anchor.js/dist/fabricators/money-market/market-register-contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricatebMarketRegisterContracts = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const fabricatebMarketRegisterContracts = ({ address, overseer_contract, market, interest_model, distribution_model, collector_contract, distributor_contract, }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, address_1.validateAddress)(overseer_contract),
    ]);
    const mmMarket = addressProvider.market(market);
    return [
        new terra_js_1.MsgExecuteContract(address, mmMarket, {
            register_contracts: {
                overseer_contract: overseer_contract,
                interest_model: interest_model,
                distribution_model: distribution_model,
                collector_contract: collector_contract,
                distributor_contract: distributor_contract,
            },
        }),
    ];
};
exports.fabricatebMarketRegisterContracts = fabricatebMarketRegisterContracts;
