// @anchor-protocol/anchor.js/dist/fabricators/money-market/market-redeem-stable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateMarketRedeemStable = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const address_1 = require("../../utils/validation/address");
const validate_input_1 = require("../../utils/validate-input");
const number_1 = require("../../utils/validation/number");
const create_hook_msg_1 = require("../../utils/cw20/create-hook-msg");
/**
 * @param address Client’s Terra address.
 * @param symbol Symbol of stablecoin to redeem, or its aToken equivalent.
 * @param amount Amount of a stablecoin to redeem, or amount of an aToken (aTerra) to redeem (specified by market).
 */
const fabricateMarketRedeemStable = ({ address, market, amount }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, number_1.validateIsGreaterThanZero)(amount),
    ]);
    const marketAddress = addressProvider.market(market);
    const aTokenAddress = addressProvider.aTerra(market);
    return [
        new terra_js_1.MsgExecuteContract(address, aTokenAddress, {
            send: {
                contract: marketAddress,
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: (0, create_hook_msg_1.createHookMsg)({
                    redeem_stable: {},
                }),
            },
        }),
    ];
};
exports.fabricateMarketRedeemStable = fabricateMarketRedeemStable;
