// @anchor-protocol/anchor.js/dist/fabricators/money-market/overseer-update-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateOverseerUpdateConfig = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const true_1 = require("../../utils/validation/true");
const number_1 = require("../../utils/validation/number");
const fabricateOverseerUpdateConfig = ({ address, overseer, owner_addr, oracle_contract, liquidation_contract, threshold_deposit_rate, target_deposit_rate, buffer_distribution_factor, anc_purchase_factor, epoch_period, price_timeframe, }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        owner_addr ? (0, address_1.validateAddress)(owner_addr) : true_1.validateTrue,
        oracle_contract ? (0, address_1.validateAddress)(oracle_contract) : true_1.validateTrue,
        liquidation_contract
            ? (0, address_1.validateAddress)(liquidation_contract)
            : true_1.validateTrue,
        epoch_period ? (0, number_1.validateIsNumber)(epoch_period) : true_1.validateTrue,
        epoch_period ? (0, number_1.validateIsNumber)(epoch_period) : true_1.validateTrue,
    ]);
    const mmOverseer = addressProvider.overseer(overseer);
    return [
        new terra_js_1.MsgExecuteContract(address, mmOverseer, {
            update_config: {
                owner_addr: owner_addr,
                oracle_contract: oracle_contract,
                liquidation_contract: liquidation_contract,
                threshold_deposit_rate: threshold_deposit_rate,
                target_deposit_rate: target_deposit_rate,
                buffer_distribution_factor: buffer_distribution_factor,
                anc_purchase_factor: anc_purchase_factor,
                epoch_period: epoch_period,
                price_timeframe: price_timeframe,
            },
        }),
    ];
};
exports.fabricateOverseerUpdateConfig = fabricateOverseerUpdateConfig;
