// @anchor-protocol/anchor.js/dist/fabricators/money-market/overseer-whitelist.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateOverseerWhitelist = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const fabricateOverseerWhitelist = ({ address, overseer, name, symbol, collateral_token, custody_contract, max_ltv, }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, address_1.validateAddress)(collateral_token),
        (0, address_1.validateAddress)(custody_contract),
    ]);
    const mmOverseer = addressProvider.overseer(overseer);
    return [
        new terra_js_1.MsgExecuteContract(address, mmOverseer, {
            whitelist: {
                name: name,
                symbol: symbol,
                collateral_token: collateral_token,
                custody_contract: custody_contract,
                max_ltv: max_ltv,
            },
        }),
    ];
};
exports.fabricateOverseerWhitelist = fabricateOverseerWhitelist;
