// @anchor-protocol/anchor.js/dist/fabricators/money-market/provide-collateral.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateProvideCollateral = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const create_hook_msg_1 = require("../../utils/cw20/create-hook-msg");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const number_1 = require("../../utils/validation/number");
const provider_1 = require("../../address-provider/provider");
/**
 *
 * @param address Client’s Terra address.
 * @param market Type of stablecoin money market to deposit collateral. Currently only supports UST and KRT.
 //  * @param borrower (optional) — Terra address of the entity that created the loan position. If null, adds collateral to address‘s loan position.
 * @param loan_id ID of address’s loan position to add collateral. For each addresses, their first loan position is given ID = 0, second loan ID = 1, third ID = 2, etc.. If this field is [(address’s current highest loan_id) + 1], a new loan position is created.
 * @param symbol Symbol of collateral to deposit.
 * @param amount Amount of collateral to deposit.
 */
const fabricateProvideCollateral = ({ address, market, collateral, amount }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        // borrower ? validateAddress(borrower) : validateTrue,
        (0, number_1.validateIsGreaterThanZero)(amount),
    ]);
    let bAssetTokenContract = addressProvider.bLunaToken();
    if (collateral == provider_1.COLLATERAL_DENOMS.UBETH) {
        bAssetTokenContract = addressProvider.bEthToken();
    }
    const mmOverseerContract = addressProvider.overseer(market);
    const custodyContract = addressProvider.custody(market, collateral);
    // cw20 send + provide_collateral hook
    /* eslint-disable */
    return [
        // provide_collateral call
        new terra_js_1.MsgExecuteContract(address, bAssetTokenContract, {
            send: {
                contract: custodyContract,
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: (0, create_hook_msg_1.createHookMsg)({
                    deposit_collateral: {},
                }),
            },
        }),
        // lock_collateral call
        new terra_js_1.MsgExecuteContract(address, mmOverseerContract, {
            // @see https://github.com/Anchor-Protocol/money-market-contracts/blob/master/contracts/overseer/src/msg.rs#L75
            lock_collateral: {
                collaterals: [
                    [
                        bAssetTokenContract,
                        new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                    ],
                ],
            },
        }),
    ];
};
exports.fabricateProvideCollateral = fabricateProvideCollateral;
