// @anchor-protocol/anchor.js/dist/fabricators/beth/beth-send-from.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricatebEthSendFrom = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const number_1 = require("../../utils/validation/number");
const fabricatebEthSendFrom = ({ address, amount, contract, owner, msg }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([
        (0, address_1.validateAddress)(address),
        (0, number_1.validateIsNumber)(amount),
        (0, number_1.validateIsGreaterThanZero)(amount),
        (0, address_1.validateAddress)(owner),
        (0, address_1.validateAddress)(contract),
    ]);
    const bEthTokenAddress = addressProvider.bEthToken();
    let message = undefined;
    if (msg) {
        message = Buffer.from(JSON.stringify(msg)).toString('base64');
    }
    return [
        new terra_js_1.MsgExecuteContract(address, bEthTokenAddress, {
            send_from: {
                owner: owner,
                contract: contract,
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: message,
            },
        }),
    ];
};
exports.fabricatebEthSendFrom = fabricatebEthSendFrom;
