// @anchor-protocol/anchor.js/dist/fabricators/anchor-token/gov-stake-voting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateGovStakeVoting = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const create_hook_msg_1 = require("../../utils/cw20/create-hook-msg");
const fabricateGovStakeVoting = ({ address, amount }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([(0, address_1.validateAddress)(address)]);
    const anchorToken = addressProvider.ANC();
    return [
        new terra_js_1.MsgExecuteContract(address, anchorToken, {
            send: {
                contract: addressProvider.gov(),
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: (0, create_hook_msg_1.createHookMsg)({
                    stake_voting_tokens: {},
                }),
            },
        }),
    ];
};
exports.fabricateGovStakeVoting = fabricateGovStakeVoting;
