// @anchor-protocol/anchor.js/dist/fabricators/anchor-token/staking-bond.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fabricateStakingBond = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const validate_input_1 = require("../../utils/validate-input");
const address_1 = require("../../utils/validation/address");
const create_hook_msg_1 = require("../../utils/cw20/create-hook-msg");
const fabricateStakingBond = ({ address, amount }) => (addressProvider) => {
    (0, validate_input_1.validateInput)([(0, address_1.validateAddress)(address)]);
    const lpToken = addressProvider.ancUstLPToken();
    return [
        new terra_js_1.MsgExecuteContract(address, lpToken, {
            send: {
                contract: addressProvider.staking(),
                amount: new terra_js_1.Int(new terra_js_1.Dec(amount).mul(1000000)).toString(),
                msg: (0, create_hook_msg_1.createHookMsg)({
                    bond: {},
                }),
            },
        }),
    ];
};
exports.fabricateStakingBond = fabricateStakingBond;
