// @anchor-protocol/anchor.js/dist/address-provider/from-json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressProviderFromJson = void 0;
const provider_1 = require("./provider");
class AddressProviderFromJson {
    constructor(data) {
        this.data = data;
    }
    bLunaReward() {
        return this.data.bLunaReward;
    }
    bLunaHub() {
        return this.data.bLunaHub;
    }
    bLunaToken() {
        return this.data.bLunaToken;
    }
    bEthReward() {
        return this.data.bEthReward;
    }
    bLunaValidatorsRegistry() {
        return this.data.bLunaValidatorsRegistry;
    }
    bEthToken() {
        return this.data.bEthToken;
    }
    market() {
        return this.data.mmMarket;
    }
    custody(_denom, collateral) {
        switch (collateral) {
            case provider_1.COLLATERAL_DENOMS.UBLUNA: {
                return this.data.mmCustody;
            }
            case provider_1.COLLATERAL_DENOMS.UBETH: {
                return this.data.mmCustodyBEth;
            }
        }
    }
    overseer() {
        return this.data.mmOverseer;
    }
    aTerra() {
        return this.data.aTerra;
    }
    oracle() {
        return this.data.mmOracle;
    }
    interest() {
        return this.data.mmInterestModel;
    }
    distribution() {
        return this.data.mmDistributionModel;
    }
    liquidation() {
        return this.data.mmLiquidation;
    }
    liquidationQueue() {
        return this.data.mmLiquidationQueue;
    }
    bLunaLunaPair() {
        return this.data.bLunaLunaPair;
    }
    bLunaLunaLPToken() {
        return this.data.bLunaLunaLPToken;
    }
    gov() {
        return this.data.gov;
    }
    ancUstPair() {
        return this.data.ancUstPair;
    }
    ancUstLPToken() {
        return this.data.ancUstLPToken;
    }
    collector() {
        return this.data.collector;
    }
    staking() {
        return this.data.staking;
    }
    community() {
        return this.data.community;
    }
    distributor() {
        return this.data.distributor;
    }
    ANC() {
        return this.data.ANC;
    }
    airdrop() {
        return this.data.airdrop;
    }
    investorLock() {
        return this.data.investor_vesting;
    }
    teamLock() {
        return this.data.team_vesting;
    }
}
exports.AddressProviderFromJson = AddressProviderFromJson;
